### PR TITLE
Install MySQL and PG Deps for Rust when Postgres and MySQL are used

### PR DESCRIPTION
**Pull request by JakeCooper**
This PR installs dependencies depending on the detected library

### CODE REVIEW

File: examples/rust-diesel/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+use std::{env, process::Command};
```
Comment (devtomio) [inline examples/rust-diesel/src/main.rs:1]: Add `rust-diesel` to the `IGNORE` constant in `test-helper`.

**Review by devtomio — CHANGES_REQUESTED** [review]

File: src/providers/rust.rs [hunk 2]
```diff
@@ -227,40 +236,29 @@ impl RustProvider {
...
-        if app.includes_file("Cargo.lock") && app.read_file("Cargo.lock")?.contains("openssl") {
-            return Ok(true);
-        }
-
-        Ok(false)
+    fn uses_dependency(app: &App, dep: &'static str) -> bool {
+        // Check if any of the files includes the dependency
+        ["Cargo.toml", "Cargo.lock"].iter().any(|file_name| {
+            app.includes_file(file_name)
+                && app.read_file(file_name).unwrap_or_default().contains(dep)
```
Comment (nebulatgs) [inline src/providers/rust.rs:260]: i don't think we should use `unwrap_or_default` here, if we fail to read the file, something has gone wrong and the error should be handled.

**Comment by JakeCooper** [comment]
Seems to be failing for some reason and can't tell

```
------                                                                                                               
 > [stage-0  9/10] RUN --mount=type=cache,id=9UgkYzoS89g-/root/cargo/git,target=/root/.cargo/git --mount=type=cache,id=9UgkYzoS89g-/root/cargo/registry,target=/root/.cargo/registry --mount=type=cache,id=9UgkYzoS89g-target,target=target cargo build --release --target aarch64-unknown-linux-musl:                                                         
#16 0.228     Updating crates.io index                                                                               
#16 1.180    Compiling rust-diesel v0.1.0 (/app)                                                                     
#16 1.267 warning: unused import: `process::Command`
#16 1.267  --> src/main.rs:1:16
#16 1.267   |
#16 1.267 1 | use std::{env, process::Command};
#16 1.267   |                ^^^^^^^^^^^^^^^^
#16 1.267   |
#16 1.267   = note: `#[warn(unused_imports)]` on by default
#16 1.267 
#16 2.401 error: linking with `cc` failed: exit status: 1
#16 2.401   |
#16 2.401   = note: "cc" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crt1.o" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crti.o" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtbegin.o" "/tmp/rustc9hH4xJ/symbols.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.0.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.1.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.10.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.11.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.2.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.3.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.4.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.5.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.6.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.7.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.8.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.rust_diesel.67e3622e-cgu.9.rcgu.o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7.3sbjddl50pmbyspt.rcgu.o" "-Wl,--as-needed" "-L" "/app/target/aarch64-unknown-linux-musl/release/deps" "-L" "/app/target/release/deps" "-L" "/usr/lib/aarch64-linux-gnu" "-L" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib" "-Wl,-Bstatic" "/app/target/aarch64-unknown-linux-musl/release/deps/libdiesel-2dfb3d0707dfde2f.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libpq_sys-3d649afea44a8e81.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libbyteorder-9ff3ec6741dea488.rlib" "/app/target/aarch64-unknown-linux-musl/release/deps/libbitflags-e8ef1c3e6c8e7295.rlib" "-Wl,--start-group" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-9a6bce422f0000ee.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libpanic_unwind-159df072af691315.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libobject-384907cca4f5909b.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libmemchr-c4cdcfa83263bff2.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libaddr2line-5ea6f9ba4a8a1cae.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libgimli-33c8299f0e34415c.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_demangle-9f6d554162f3daa1.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd_detect-f233b509811a3b1c.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libhashbrown-9bad9e9ee53588e5.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libminiz_oxide-0aaba3d269092688.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libadler-84369741afc56366.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_alloc-a771e22077bae3b9.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libunwind-a4f4d8b1097e894a.rlib" "-lunwind" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcfg_if-8a591157c24046a9.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liblibc-7171e8d387a44fc7.rlib" "-lc" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/liballoc-df8b9962134a270a.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/librustc_std_workspace_core-fab6effc05523a23.rlib" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcore-3fe103e3cb247495.rlib" "-Wl,--end-group" "/nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libcompiler_builtins-413a2b918f1f8443.rlib" "-Wl,-Bdynamic" "-lpq" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-nostartfiles" "-L" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib" "-L" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained" "-o" "/app/target/aarch64-unknown-linux-musl/release/deps/rust_diesel-0d08aa6112be4ce7" "-Wl,--gc-sections" "-static" "-no-pie" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtend.o" "/nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/crtn.o"
#16 2.401   = note: /nix/store/4m0ql1nd3kahkcj9ah1s78pc689mfmhf-binutils-2.38/bin/ld: /nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-9a6bce422f0000ee.rlib(std-9a6bce422f0000ee.std.3f650011-cgu.0.rcgu.o): in function `std::sys::unix::os::home_dir::fallback':
#16 2.401           /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys/unix/os.rs:627: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#16 2.401           /nix/store/4m0ql1nd3kahkcj9ah1s78pc689mfmhf-binutils-2.38/bin/ld: /nix/store/131iqai06w79dc8pxfz3hndxqrqhdq6l-rust-std-1.63.0-aarch64-unknown-linux-musl/lib/rustlib/aarch64-unknown-linux-musl/lib/libstd-9a6bce422f0000ee.rlib(std-9a6bce422f0000ee.std.3f650011-cgu.0.rcgu.o): in function `<std::sys_common::net::LookupHost as core::convert::TryFrom<(&str,u16)>>::try_from':
#16 2.401           /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/sys_common/net.rs:205: warning: Using 'getaddrinfo' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
#16 2.401           /nix/store/4m0ql1nd3kahkcj9ah1s78pc689mfmhf-binutils-2.38/bin/ld: /nix/store/xywnx0f0fcc3wc69mr4zr298yk97pw3l-rust-default-1.63.0/lib/rustlib/aarch64-unknown-linux-musl/lib/self-contained/libunwind.a(libunwind.o): undefined reference to symbol 'dladdr@@GLIBC_2.34'
#16 2.401           /nix/store/4m0ql1nd3kahkcj9ah1s78pc689mfmhf-binutils-2.38/bin/ld: /nix/store/il2mpl6qjc4aa42mbiy3p3wicm3d6nz5-glibc-2.35-163/lib/libc.so.6: error adding symbols: DSO missing from command line
#16 2.401           collect2: error: ld returned 1 exit status
#16 2.401           
#16 2.401   = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
#16 2.401   = note: use the `-l` flag to specify native libraries to link
#16 2.401   = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)
#16 2.401 
#16 2.412 warning: `rust-diesel` (bin "rust-diesel") generated 1 warning
#16 2.412 error: could not compile `rust-diesel` due to previous error; 1 warning emitted
------
executor failed running [/bin/bash -ol pipefail -c cargo build --release --target aarch64-unknown-linux-musl]: exit code: 101
Error: Docker build failed
```

**Comment by JakeCooper** [comment]
Get farther with NO_MUSL

```
     Running `target/debug/nixpacks build ./examples/rust-diesel-postgres --env NO_MUSL=false`

╔═══════════════════════════ Nixpacks v0.3.4 ═══════════════════════════╗
║ Setup      │ gcc, rust-bin.stable.latest.default, libpq-dev, binutils ║
║───────────────────────────────────────────────────────────────────────║
║ Install    │                                                          ║
║───────────────────────────────────────────────────────────────────────║
║ Build      │ mkdir -p bin                                             ║
║            │ cargo build --release                                    ║
║            │ cp target/release/rust-diesel bin                        ║
║───────────────────────────────────────────────────────────────────────║
║ Start      │ ./bin/rust-diesel                                        ║
╚═══════════════════════════════════════════════════════════════════════╝

[+] Building 0.6s (15/15) FINISHED                                                                                                                                                                                                                                                                
 => [internal] load build definition from Dockerfile                                                                                                                                                                                                                                         0.1s
 => => transferring dockerfile: 1.17kB                                                                                                                                                                                                                                                       0.0s
 => [internal] load .dockerignore                                                                                                                                                                                                                                                            0.0s
 => => transferring context: 2B                                                                                                                                                                                                                                                              0.0s
 => [internal] load metadata for ghcr.io/railwayapp/nixpacks:debian-1660071566                                                                                                                                                                                                               0.3s
 => [stage-0  1/10] FROM ghcr.io/railwayapp/nixpacks:debian-1660071566@sha256:aa5a574d94cd1c72b367921d0dd6b586b89490f81ada2e6cfba352bf300b1026                                                                                                                                               0.0s
 => [internal] load build context                                                                                                                                                                                                                                                            0.0s
 => => transferring context: 2.82kB                                                                                                                                                                                                                                                          0.0s
 => CACHED [stage-0  2/10] WORKDIR /app/                                                                                                                                                                                                                                                     0.0s
 => CACHED [stage-0  3/10] COPY .nixpacks/setup.nix .nixpacks/setup.nix                                                                                                                                                                                                                      0.0s
 => CACHED [stage-0  4/10] RUN nix-env -if .nixpacks/setup.nix                                                                                                                                                                                                                               0.0s
 => CACHED [stage-0  5/10] RUN apt-get update && apt-get install -y libpq-dev binutils                                                                                                                                                                                                       0.0s
 => CACHED [stage-0  6/10] COPY . /app/                                                                                                                                                                                                                                                      0.0s
 => CACHED [stage-0  7/10] COPY . /app/                                                                                                                                                                                                                                                      0.0s
 => CACHED [stage-0  8/10] RUN --mount=type=cache,id=9UgkYzoS89g-/root/cargo/git,target=/root/.cargo/git --mount=type=cache,id=9UgkYzoS89g-/root/cargo/registry,target=/root/.cargo/registry --mount=type=cache,id=9UgkYzoS89g-target,target=target mkdir -p bin                             0.0s
 => CACHED [stage-0  9/10] RUN --mount=type=cache,id=9UgkYzoS89g-/root/cargo/git,target=/root/.cargo/git --mount=type=cache,id=9UgkYzoS89g-/root/cargo/registry,target=/root/.cargo/registry --mount=type=cache,id=9UgkYzoS89g-target,target=target cargo build --release                    0.0s
 => CACHED [stage-0 10/10] RUN --mount=type=cache,id=9UgkYzoS89g-/root/cargo/git,target=/root/.cargo/git --mount=type=cache,id=9UgkYzoS89g-/root/cargo/registry,target=/root/.cargo/registry --mount=type=cache,id=9UgkYzoS89g-target,target=target cp target/release/rust-diesel bin        0.0s
 => exporting to image                                                                                                                                                                                                                                                                       0.0s
 => => exporting layers                                                                                                                                                                                                                                                                      0.0s
 => => writing image sha256:427ac0ec5b8ee348a530049e18cb4d1ee3d85ed99ee82e49821f9b69cce5b08d                                                                                                                                                                                                 0.0s
 => => naming to docker.io/library/5cd3156f-a221-4557-864f-606f62e251c9                                                                                                                                                                                                                      0.0s

Use 'docker scan' to run Snyk tests against images to find vulnerabilities and learn how to fix them
=== Successfully Built! ===

Run:
  docker run -it 5cd3156f-a221-4557-864f-606f62e251c9
➜  nixpacks git:(cooper/diesel) ✗ docker run -it 5cd3156f-a221-4557-864f-606f62e251c9
./bin/rust-diesel: error while loading shared libraries: libssl.so.1.1: cannot open shared object file: No such file or directory
```

CC @nebulatgs

**Comment by JakeCooper** [comment]
By adding NIXPKGS=openssl,openssl.dev,libkrb5 with NO_MUSL

I managed to get it to ask for `./bin/rust-diesel: error while loading shared libraries: libldap_r-2.4.so.2: cannot open shared object file: No such file or directory`

If you run it with musl, it gets stuck on `./rust-diesel: error while loading shared libraries: libpq.so.5: cannot open shared object file: No such file or directory`

Dunno what to do here so probably gonna have to just close this out

**Comment by JakeCooper** [comment]
CC @FinnDore if you wanna take a stab

**Comment by coffee-cup** [comment]
> If you run it with musl, it gets stuck on ./rust-diesel: error while loading shared libraries: libpq.so.5: cannot open shared object file: No such file or directory

This is likely the case because with MUSL enabled the binary is copied to a new debian slim image where the shared header files will not exist

**Comment by FinnDore** [comment]
Thanks for looking into this @JakeCooper 🙏, I wish dep battleships were [this](https://www.youtube.com/watch?v=ectnFkHyeEs) easy.

Ive just tried tryed to translate [this](https://github.com/FinnDore/something/blob/4c28ab65ea5999a70dc458577007dea3de40b005/apps/something/dockerfile)  ( or the relevant parts ) into the rust provider + editing the `generate_dockerfile` function. But no luck, Ill continue to look at this in the background but im really waiting for a rust god to step in. It really does feel like this library is built on a stack of paper cards lol.

Interestingly in my docker image i do copy the built binary, but it just works when i re-install the deps in the `debian:buster-slim` container. Also notice how i don't need to install the deps in the first build container. Odd!

The docker file that works:
```dockerfile
ARG ROCKET_ADDRESS=0.0.0.0
ARG DATABASE_URL
ARG AUTH_SECRET

FROM amd64/rust:1.62.10@sha256:4f1d43c216d995c2f734d7c682bc3e2abe3e110961cda4ae0743ce3944e673a2 as builder

WORKDIR /usr/src/something
COPY . .

EXPOSE 80

RUN cargo build --release

FROM debian:buster-slim

COPY --from=builder /usr/src/something/target/release/something /bin
COPY --from=builder /usr/src/something/apps/something/rocket.toml /usr

RUN apt-get update && apt-get install -y libssl-dev ca-certificates libmariadb-dev && rm -rf /var/lib/apt/lists/*

ENV AUTH_SECRET=${AUTH_SECRET}
ENV ROCKET_ADDRESS=${ROCKET_ADDRESS}
ENV DATABASE_URL=${DATABASE_URL}

ENV ROCKET_CONFIG=/usr/rocket.toml

CMD ["something"]
```

**Comment by JakeCooper** [comment]
> Interestingly in my docker image i do copy the built binary, but it just works when i re-install the deps in the debian:buster-slim

I assume because the build actually happens in that rust container which, at some point, someone added the correct header files for :/

**Pull request closed without merging**